File: src/libraryclient/uidataprovider.cpp
/*
 * niepce - libraryclient/uidataprovider.cpp
 *
 * Copyright (C) 2011-2022 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <algorithm>

#include "fwk/base/debug.hpp"
#include "uidataprovider.hpp"

namespace libraryclient {

void UIDataProvider::updateLabel(const eng::LabelPtr& l)
{
    // TODO: will work as long as we have 5 labels or something.
    for (auto & label : m_labels) {
        if (label->id() == l->id()) {
            label = l->clone_boxed();
        }
    }
}


void UIDataProvider::addLabel(const eng::LabelPtr& l)
{
    m_labels.push_back(l->clone_boxed());
}


void UIDataProvider::deleteLabel(int id)
{
    // TODO: will work as long as we have 5 labels or something.
    for(auto iter = m_labels.begin();
        iter != m_labels.end(); ++iter) {

        if ((*iter)->id() == id) {
            DBG_OUT("remove label %d", id);
            iter = m_labels.erase(iter);
            break;
        }
    }
}

std::optional<fwk::RgbColourPtr> UIDataProvider::colourForLabel(int32_t id) const
{
    for(auto& label : m_labels) {
        if (label->id() == id) {
            return std::optional<fwk::RgbColourPtr>(label->colour());
        }
    }
    return std::nullopt;
}


}

/*
  Local Variables:
  mode:c++
  c-file-style:"stroustrup"
  c-file-offsets:((innamespace . 0))
  indent-tabs-mode:nil
  fill-column:80
  End:
*/

File: crates/npc-engine/src/libraryclient/ui_data_provider.rs
/*
 * niepce - libraryclient/uidataprovider.rs
 *
 * Copyright (C) 2022-2025 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::cell::RefCell;

use crate::catalog::{Label, LibraryId};
use npc_fwk::base::rgbcolour::RgbColour;

/// UIDataProvider provide data for the library for the UI
/// Currently handle the `Label`
pub struct UIDataProvider {
    labels: RefCell<Vec<Label>>,
}

impl Default for UIDataProvider {
    fn default() -> UIDataProvider {
        UIDataProvider {
            labels: RefCell::new(Vec::default()),
        }
    }
}

impl UIDataProvider {
    pub fn add_label(&self, label: &Label) {
        self.labels.borrow_mut().push(label.clone());
    }

    pub fn update_label(&self, label: &Label) {
        let id = label.id();
        for l in &mut self.labels.borrow_mut().iter_mut() {
            if l.id() == id {
                *l = label.clone();
                break;
            }
        }
    }

    pub fn delete_label(&self, label: LibraryId) {
        let mut labels = self.labels.borrow_mut();
        if let Some(idx) = labels.iter().position(|l| l.id() == label) {
            labels.remove(idx);
        }
    }

    pub fn colour_for_label(&self, id: LibraryId) -> RgbColour {
        if let Some(label) = self.labels.borrow().iter().find(|l| l.id() == id) {
            label.colour().clone()
        } else {
            RgbColour::default()
        }
    }

    pub fn label_count(&self) -> usize {
        self.labels.borrow().len()
    }

    pub fn label_at(&self, idx: usize) -> Label {
        self.labels.borrow()[idx].clone()
    }
}
